### PR TITLE
UI add an artifact view on notebooks

### PR DESCRIPTION
**Pull request by nlaille**
## What changes are proposed in this pull request?
 
Add an artifact view for notebooks
 
## How is this patch tested?
 
Manually.
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Add a view for notebook artifact in mlflow ui.
 
### What component(s) does this PR affect?
 
- [x] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by mateiz** [comment]
Thanks for finding this -- using nteract for this is a great idea. I had a few questions though:

1) How does nteract isolate notebook content from the rest of the application from a security perspective? Is it meant to? For example, if I load a notebook artifact that is displaying a cell with JavaScript, can that cell access stuff on the whole page, or is it treated as being from the same origin as the MLflow UI?

2) What is the state of compatibility between nteract and Jupyter's own UIs (Jupyter and JupyterLab)? Those are used more often than nteract as far as I know.

3) Would it be possible to just use Jupyter's display components instead -- does it offer such a thing?

**Comment by nlaille** [comment]
@mateiz : 

1. I will run some tests to be sure, but the code should not be executed. By the way, `notebook-preview` is deprecated so I am going to move to `presentational-components`. 
2. nteract has his own desktop application now. But it is still available through a jupyter extension. And they are both using the same syntax for ipynb.
3. I was not able to find some javascript components from jupyter to render notebooks at client side.

**Comment by nlaille** [comment]
@mateiz , for the first point : `presentational-components` is just used for rendering and `react` is escaping string so it should prevent it. But I am not an expert on it, so feedbacks from someone else are welcome.

**Comment by mateiz** [comment]
OK, we'll definitely need someone to look at the security implications if these things can run JavaScript or more generally make arbitrary HTTP requests with the same origin as our page. For 2 and 3, do you know whether there are limitations in the output types that nteract can display compared to Jupyter? For example, can it also display images, HTML outputs, widgets, etc?

**Comment by nlaille** [comment]
It works for all kind of content : https://components.nteract.io/#section-nteractoutputsmedia 
I tested it with standard output, error, image... But I can do a test with a more complex notebook to be sure.
There is some issues with css on standard output and error because of code.less stylesheet but it is looking nice !

**Comment by Zangr** [comment]
@nlaille, could you please provide a gif/screenshot of the view implemented and try rebasing to resolve the conflicting files?

**Comment by nlaille** [comment]
image
image
image

Some css might be add to override default css for outputs.

**Comment by mateiz** [comment]
This definitely looks cool! We should investigate the possible XSS/CORS implications and get this in if we can avoid those.

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 1]
```diff
@@ -0,0 +1,210 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { getSrc } from './ShowArtifactPage';
+import './ShowArtifactNbView.css';
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: Let's put CSS imports after js imports

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    LightTheme, Cell, Input, Prompt, Source, Outputs, Cells
+} from "@nteract/presentational-components";
+import {
+    Output, StreamText, DisplayData, Media, KernelOutputError, ExecuteResult
+} from "@nteract/outputs";
+import Markdown from "@nteract/markdown";
+import { Provider as MathJaxProvider } from "@nteract/mathjax";
+
+
+class ShowArtifactNbView extends Component {
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: What about `ArtifactNotebookView`?

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+  };
+
+  state = {
+    loading: true,
+    error: undefined,
+    immutableNotebook: undefined
+  };
+
+  componentWillMount() {
+    this.fetchArtifacts();
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: `componentWillMount` is deprecated in React 16 and will be removed in 17. Let's do the initial fetching in `componentDidMount`.

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+      headers: new Headers(getRequestHeaders(document.cookie))
+    });
+    fetch(getArtifactRequest).then((response) => {
+      return response.blob();
+    }).then((blob) => {
+      const fileReader = new FileReader();
+      fileReader.onload = (event) => {
+        const notebook = parseNotebook(event.target.result);
+        const immutableNotebook = fromJS(notebook);
+        this.setState({ immutableNotebook: immutableNotebook, loading: false });
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: Should we reset error after loading succeed here?

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+      };
+      fileReader.readAsText(blob);
+    });
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: Should we have a catch block here to set error state?

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+  render() {
+    if (this.state.loading) {
+      return (
+        <div>
+          Loading...
+        </div>
+      );
+    }
+    if (this.state.error) {
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: I'd recommend evaluating the state in the following way:
```
const { error, loading, immutableNotebook } = this.state;
if (error) { ... } else if(loading) { ... } else if (immutableNotebook) { ... } else { ... }
```

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                                  `}</style>
+                                </Cell>
+                              );
+                            case "raw":
+                              return (
+                                <Cell key={cellID}>
+                                  <pre className="raw-cell">
+                                    {source}
+                                    <style jsx>{`
+                                      raw-cell {
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: Can we move this to the CSS file?

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+                                </Cell>
+                              );
+                            case "markdown":
+                              return (
+                                <Cell key={cellID}>
+                                  <div className="content-margin">
+                                    <Markdown source={source} />
+                                  </div>
+                                  <style jsx>{`
+                                    .content-margin {
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: Can we move this to the CSS file?

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+                                  <Outputs>
+                                    <pre>{`Cell Type "${cellType}" is not implemented`}</pre>
+                                  </Outputs>
+                                </Cell>
+                              );
+                          }
+                        })}
+                    </Cells>
+
+                    <style>{`:root {
```
Comment (Zangr) [inline mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: Can we move this to the CSS file?

**Review by Zangr — COMMENTED** [review]
Nice work! I went through one run and left some comments. Could you please also provide some basic rendering tests?

File: mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js (hunk 2 above)
Comment (nlaille) [reply mlflow/server/js/src/components/artifact-view-components/ShowArtifactNbView.js]: I wanted to keep the same syntax as `ShowArtifactTextView` and `ShowArtifactImageView`. Not sure to make this change in this pr, but I prefer `NotebookArtifactView`.

**Comment by nlaille** [comment]
> Nice work! I went through one run and left some comments. Could you please also provide some basic rendering tests?

@Zangr I never done a rendering test in react. And this one is kind of tricky because of async rendering. Should I refactor `ShowArtifactNbView` to make it testable or do you have some tests example I can use to do it ?

**Comment by Zangr** [comment]
@nlaille thanks for adding the test. Implementation of this PR looks generally good! 

One thing that is kind of blocking at the moment is that when I tried to `npm run build`, I found the nteract component introduced blocking the build with the following error. 
```
Failed to compile.

Failed to minify the code from this file:

 	./node_modules/@nteract/commutable/lib/primitives.js:31
```
Could you please take a look at it?

**Comment by nlaille** [comment]
@Zangr , I created an issue in nteract repository : https://github.com/nteract/nteract/issues/4513, I hope they can help me to fix

**Comment by nlaille** [comment]
@Zangr I had to upgrade create-react-app to v3 : https://github.com/octokit/rest.js/issues/828.
It seems to work.

**Comment by patryk-oleniuk** [comment]
Any info on when this PR will get into master? I'd love to see it in the next release :). Thanks for help.

**Comment by Zangr** [comment]
Thanks @nlaille, the change built fine with create-react-app v3. I'll do more tests next week. Also, I noticed you bumped up versions of many dependencies and some of them are major version changes. I'd recommend to keep the dependency changes to the ones that are related to this PR.

**Comment by nlaille** [comment]
@Zangr Upgrade to react-scripts v3 has a bit more impact than I thought. I think I will create a dedicate PR for it. Does it make sense for you ?

**Comment by nlaille** [comment]
Waiting for https://github.com/mlflow/mlflow/pull/1812 to be merged.

**Comment by smurching** [comment]
@nlaille one thing that came up during a chat with some security folks - it seems the notebook rendering in `nteract` is vulnerable to XSS attacks, e.g. viewing a notebook with contents like:

```
{
  "cells": [
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {},
      "outputs": [
        {
          "metadata": {},
          "output_type": "display_data",
          "data": {
            "application/javascript": [
              "alert(document.cookie)"
            ]
          }
        }
      ],
      "execution_count": 3
    }
  ],
  "metadata": {
    "name": "test-artifact-generation",
    "notebookId": 9
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
```

causes the embedded Javascript code (`alert(document.cookies)`) to be executed by the browser. This could allow an attacker to reuse cookies to make API requests (delete runs/experiments etc), even if the user is running their tracking server behind authentication.

One way around this might be to find a way to render the component inside an iframe with the [sandbox attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe) set to `"allow-scripts"` (basically, allow running JS within the iframe, but treat that JS as running from a separate origin so it can't reuse cookies from the parent page).

For example, we could consider doing so via a library like [this one](https://github.com/ryanseddon/react-frame-component) (although it seems that particular library, `react-frame-component`, [doesn't play well with the `sandbox` attribute](https://github.com/ryanseddon/react-frame-component/issues/115)). We might also be able to implement this by directly invoking the [`ReactDOM.render` method](https://reactjs.org/docs/react-dom.html#render) to inject the component into an iframe, or maybe calling `ReactDomServer.renderToString` and passing the generated HTML as the `src` attribute of an iframe - see e.g. [this blog post](https://medium.com/@ryanseddon/rendering-to-iframes-in-react-d1cb92274f86) for reference. If you're able to look into these options that'd be awesome, let me know if/what you find anything - thanks!

**Comment by mateiz** [comment]
Maybe nteract also has some options to safely sandbox content? It would be good to look into that too just in case.

**Comment by nlaille** [comment]
I take a look at nteract Media.Javascript : https://github.com/nteract/nteract/blob/master/packages/outputs/src/components/media/javascript.md 

> Under the hood, this component executes the JavaScript string passed to data using eval. As such, be sure that whatever data you are passing to the component is coming from a trusted source.

So, I will remove Media.Javascript component and add a custom one.

Also, about HTML components, nteract is using [Range Object](https://developer.mozilla.org/en-US/docs/Web/API/Range) : https://github.com/nteract/nteract/blob/master/packages/outputs/src/components/media/html.md
I don't know much about this object, so I will run some tests.

**Comment by nlaille** [comment]
As a first step I will remove Media.HTML and Media.Javascript, this way, there should not have any more issues and we will be able to merge it.
When I will have more times I could take a look to implement something about it.

If it is okay with you (@smurching and @Zangr), I will resolve conflicts and we will have to merge : https://github.com/mlflow/mlflow/pull/1812 and https://github.com/mlflow/mlflow/pull/1869 to move forward.

**Comment by nlaille** [comment]
Hi @smurching and @Zangr, I will have some spare time in the upcoming weeks.

If I am working on it, is it possible to finish it ? Or is it too much work ? (https://github.com/mlflow/mlflow/pull/1812, https://github.com/mlflow/mlflow/pull/1869)

**Comment by smurching** [comment]
Hi @nlaille, apologies for the delay! We're happy to review & continue with these PRs if you're still interested in doing so - also, happy to discuss further on the [MLflow Slack channel](https://join.slack.com/t/mlflow-users/shared_invite/enQtMzkxMTAwNTcyODM5LTkzMDFhNzliNjExOGQ1ZGI1ZmFlMGE5YWE1OTI4ZGM1ZWZmYzc3NGNiZTM3YjgwOTdlODAzMjJhZTdiN2Y3MWY) (my username is "sid") for faster iteration & feedback

**Comment by smurching** [comment]
Excluding the Media.HTML & Media.Javascript components seems like a reasonable initial step - we should also audit the other Media components to ensure there aren't others that could lead to XSS injection

**Comment by Zangr** [comment]
There are some large changes we plan to merge this week including #2251 and some extra miscellaneous. We can try upgrading to react-scripts v3 and make sure all UTs passes and do some manual tests. I'll update this Issue after preliminary PRs are merged.

**Pull request closed without merging**